### PR TITLE
[8.0][IMP][sale_service_project] Improve module technically and functionally

### PR DESCRIPTION
**Pull request by carlosdauden**
This PR improves sale_service_project addon:
- Calculate works and materials price applying pricelist
- Add sequence to order works and materials
- Add "Detailed Time" option
- Better layout for sale order and invoice PDF report, printing prices when compute_price = True
- More test coverage

This PR replaces #282

### CODE REVIEW

**Comment by carlosdauden** [comment]
cc @antespi @sergio-teruel @pedrobaeza @rafaelbn

File: sale_service_project/models/product.py [hunk 1]
```diff
@@ -34,13 +34,15 @@ def action_compute_price(self):
...
     _name = 'product.task.work'
+    _order = 'sequence'
 
     product_id = fields.Many2one(
         comodel_name='product.template', string='Product', ondelete='restrict')
     name = fields.Char(string='Name')
     hours = fields.Float(
         string='Hours',
-        digits_compute=dp.get_precision('Product Unit of Measure'))
+        digits=dp.get_precision('Product UoS'))
```
Comment (pedrobaeza) [inline sale_service_project/models/product.py]: The name of the digits precision is 'Product Unit of Measure', so it was correct before (although the function is `digits`, not `digits_compute`)

File: sale_service_project/models/product.py [hunk 2]
```diff
@@ -63,11 +65,13 @@ def action_compute_price(self):
...
+    _order = 'sequence'
 
     product_id = fields.Many2one(
         comodel_name='product.template', string='Product', ondelete='restrict')
     material_id = fields.Many2one(
         comodel_name='product.product', string='Material', required=True)
     quantity = fields.Float(
         string='Quantity',
-        digits_compute=dp.get_precision('Product Unit of Measure'))
+        digits=dp.get_precision('Product UoS'))
```
Comment (pedrobaeza) [inline sale_service_project/models/product.py]: The same

File: sale_service_project/models/sale.py [hunk 3]
```diff
@@ -81,10 +82,27 @@ class SaleOrderLine(models.Model):
...
         domain=[('type', '=', 'service')],
         string='Work Price Product')
+    task_work_product_price = fields.Float(
+        compute='_compute_task_work_values',
+        digits=dp.get_precision('Product Price'),
+        readonly=True)
+    task_work_hours = fields.Float(
+        string='Total Hours',
+        compute='_compute_task_work_values',
+        digits=dp.get_precision('Product UoS'))
```
Comment (pedrobaeza) [inline sale_service_project/models/sale.py]: 'Product Unit of Measure'

File: sale_service_project/models/sale.py [hunk 4]
```diff
@@ -93,25 +111,48 @@ def _compute_task_closed(self):
...
+        for line in self.filtered(lambda x: x.task_materials_ids):
+            materials_amount = 0.0
+            for material in line.task_materials_ids:
+                materials_amount += material.quantity * material.price
+            line.task_materials_amount = materials_amount
+
+    @api.multi
     def product_id_change(
-        self, pricelist, product, qty=0, uom=False, qty_uos=0, uos=False,
+        self, pricelist, product_id, qty=0, uom=False, qty_uos=0, uos=False,
```
Comment (pedrobaeza) [inline sale_service_project/models/sale.py]: Don't rename this. Upstream variables are wrong, I know, but there can be people that uses all arguments with name (indeed, I remember one in Odoo), and renaming this you'll get an error.

File: sale_service_project/models/sale.py [hunk 5]
```diff
@@ -93,25 +111,48 @@ def _compute_task_closed(self):
...
+                line.task_work_hours * line.task_work_product_price)
+
+    @api.multi
+    @api.depends('task_materials_ids')
+    def _compute_task_materials_amount(self):
+        for line in self.filtered(lambda x: x.task_materials_ids):
+            materials_amount = 0.0
+            for material in line.task_materials_ids:
+                materials_amount += material.quantity * material.price
+            line.task_materials_amount = materials_amount
```
Comment (pedrobaeza) [inline sale_service_project/models/sale.py]: You can shorten this with:

```
line.task_materials_amount = sum(
    line.task_materials_ids.mapped(lambda x: x.quantity * x.price))
```

File: sale_service_project/models/sale.py [hunk 6]
```diff
@@ -125,38 +166,47 @@ def product_id_change(
...
     _name = 'sale.order.line.task.work'
+    _order = 'sequence'
 
     order_line_id = fields.Many2one(
         comodel_name='sale.order.line', string='Order Line')
     name = fields.Char(string='Name')
     hours = fields.Float(
-        string='Hours',
-        digits_compute=dp.get_precision('Product Unit of Measure'))
+        string='Hours', digits=dp.get_precision('Product UoS'))
```
Comment (pedrobaeza) [inline sale_service_project/models/sale.py]: 'Product Unit of Measure'

File: sale_service_project/models/sale.py [hunk 7]
```diff
@@ -125,38 +166,47 @@ def product_id_change(
...
+    _order = 'sequence'
 
     order_line_id = fields.Many2one(
         comodel_name='sale.order.line', string='Order Line')
     product_id = fields.Many2one(
         comodel_name='product.product', string='Material')
     quantity = fields.Float(
-        string='Quantity',
-        digits_compute=dp.get_precision('Product Unit of Measure'))
+        string='Quantity', digits=dp.get_precision('Product UoS'))
```
Comment (pedrobaeza) [inline sale_service_project/models/sale.py]: 'Product Unit of Measure'

File: sale_service_project/views/report_common.xml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+     License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html). -->
+<openerp>
+<data>
+
+<template id="minimal_layout"
+          inherit_id="report.minimal_layout"
+          name="Sale service project report CSS">
+    <xpath expr="//head" position="inside">
+        <link rel="stylesheet"
+              href="/sale_service_project/static/src/css/report.css"/>
```
Comment (pedrobaeza) [inline sale_service_project/views/report_common.xml]: With this, you're overwriting CSS for all Odoo reports. Are you sure this is what you want?

Comment (carlosdauden) [inline sale_service_project/views/report_common.xml]: In principle no, but I want apply the slyle in sale order and invoice "original" reports.
Can you give me a clue?

Comment (pedrobaeza) [inline sale_service_project/views/report_common.xml]: Well, I'm seeing that you are adding new CSS classes, so there's no need of isolating the style.

Comment (pedrobaeza) [inline sale_service_project/views/report_common.xml]: Anyway, for inserting local CSS changes, look at this comment: https://github.com/odoo/odoo/issues/4359#issuecomment-67839051

**Comment by carlosdauden** [comment]
@pedrobaeza changes done

**Comment by pedrobaeza** [comment]
:+1:

**Comment by pedrobaeza** [comment]
cc @rafaelbn @sergio-teruel

**Comment by sergio-teruel** [comment]
👍

**Comment by rafaelbn** [comment]
:+1:

**Pull request merged**